src/discordcr/mention.cr

```diff
@@ -0,0 +1,109 @@
+# Utility module for parsing mentions out of Discord messages
+module Discord::Mention
+  # A hash map of regex describing how mentions are parsed by type
+  MENTION_REGEX = {
+    User     => /<@!?(?<id>\d+)>/,
+    Role     => /<@&(?<id>\d+)>/,
+    Channel  => /<#(?<id>\d+)>/,
+    Emoji    => /<(?<animated>a)?:(?<name>\w+):(?<id>\d+)>/,
+    Everyone => /@everyone/,
+    Here     => /@here/,
+  }
+
+  module SnowflakeMention
+    getter size : Int32
+    getter position : Int32
+    getter id : UInt64
+
+    def initialize(match : Regex::MatchData)
+      @size = match.size
+      @position = match.begin.not_nil!
+      @id = match["id"].to_u64
+    end
+  end
+
+  struct User
+    include SnowflakeMention
+  end
+
+  struct Role
+    include SnowflakeMention
+  end
+
+  struct Channel
+    include SnowflakeMention
+  end
+
+  struct Emoji
+    include SnowflakeMention
+
+    getter name : String
+    getter animated : Bool
+
+    def initialize(match : Regex::MatchData)
+      super
+      @name = match["name"]
+      @animated = !match["animated"]?.nil?
+    end
+  end
+
+  module PresenceMention
+    getter position : Int32
+
+    def initialize(match : Regex::MatchData)
+      @position = match.begin.not_nil!
+    end
+  end
+
+  struct Everyone
+    include PresenceMention
+
+    def size
+      9
+    end
+  end
+
+  struct Here
+    include PresenceMention
+
+    def size
+      5
+    end
+  end
+
+  alias MentionType = User | Role | Channel | Emoji | Everyone | Here
+
+  # Returns an array of mentions found in a string
+  def self.parse(string : String)
+    results = [] of MentionType
+    MENTION_REGEX.each do |type, regexp|
```

Comment:
How TF does this work, surely you'll get the results in the wrong order? Also regexes suck please parse properly it's not actually that hard.

Reply:
It was in my last implementation and wasn't commented on, so I didn't change it. Didn't think it would matter that much for 2000 character strings either. I'll change this to a parser then.